### PR TITLE
Allow DelegatedIdentity API clients to subscribe by PID 

### PR DESCRIPTION
**Pull request by bleggett**
**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**
DelegateIdentity API

**Description of change**
This is the impl side of https://github.com/spiffe/spire-api-sdk/pull/58

~For this to build you need to locally check out the `spire-api-sdk` fork [@40d1bc186c2da087756115d9e6eb45e75aef0b3b](https://github.com/bleggett/spire-api-sdk/commit/40d1bc186c2da087756115d9e6eb45e75aef0b3b) as a sibling of this repo, so until/if that is merged, **this should not be.**~

~CI will not pass until :point_up: happens!~

Ended up being simpler than I expected.

~Note that docs and examples are pending still.~

**Which issue this PR fixes**
#5019

### CODE REVIEW

File: pkg/agent/api/delegatedidentity/v1/service.go [hunk 1]
```diff
@@ -111,6 +113,63 @@ func (s *Service) isCallerAuthorized(ctx context.Context, log logrus.FieldLogger
...
+	log := rpccontext.Logger(ctx)
+	var receivedFirstUpdate bool
+
+	cachedSelectors, err := s.isCallerAuthorized(ctx, log, nil)
+	if err != nil {
+		return err
+	}
+
+	// Delegate authorized, use PID the delegate gave us to try and attest OBO
+	selectors, err := s.delegateWorkloadAttestor.Attest(ctx, int(req.Pid))
```
Comment (bleggett) [inline pkg/agent/api/delegatedidentity/v1/service.go:134]: Note that `Attest` takes `int` here - AFAIK PIDs are never negative, so is there a reason we didn't use `uint` here

Comment (azdagron) [reply pkg/agent/api/delegatedidentity/v1/service.go:134]: There are many arguments against the use of unsigned integers. All of that aside, `int` matches the type we receive when querying the linux kernel for the caller info, e.g. pid_t type on the pid field in the ucred struct. The Go stdlib also uses int's to represent process IDs. We should continue using signed integers.

File: pkg/agent/api/delegatedidentity/v1/service.go [hunk 2]
```diff
@@ -111,6 +113,63 @@ func (s *Service) isCallerAuthorized(ctx context.Context, log logrus.FieldLogger
...
+	ctx := stream.Context()
+	log := rpccontext.Logger(ctx)
+	var receivedFirstUpdate bool
+
+	cachedSelectors, err := s.isCallerAuthorized(ctx, log, nil)
+	if err != nil {
+		return err
+	}
+
+	// Delegate authorized, use PID the delegate gave us to try and attest OBO
```
Comment (sorindumitru) [inline pkg/agent/api/delegatedidentity/v1/service.go:133]: nit: the PR doesn't mention what OBO means and it's not a common abbrevation. Might be good to just say "on-behalf-of".

**Comment by bleggett** [comment]
(WAS passing until I resynced with `main` - that `windows unit test` seems like an unrelated flake tho)

File: pkg/agent/api/delegatedidentity/v1/service.go [hunk 3]
```diff
@@ -111,6 +113,55 @@ func (s *Service) isCallerAuthorized(ctx context.Context, log logrus.FieldLogger
...
+func (s *Service) constructValidSelectorsFromReq(ctx context.Context, log logrus.FieldLogger, reqPid int32, reqSelectors []*types.Selector) ([]*common.Selector, error) {
+	// If you set
+	// - both pid and selector args
+	// - neither of them
+	// it's an error
+	// NOTE: the default value of int32 is naturally 0 in protobuf, which is also a valid PID.
+	// However, we will still treat that as an error, as we do not expect to ever be asked to attest
+	// pid 0.
+
+	fmt.Printf("PIDS: %d", reqPid)
```
Comment (azdagron) [inline pkg/agent/api/delegatedidentity/v1/service.go:125]: Leftover debug printf?

File: pkg/agent/api/delegatedidentity/v1/service.go [hunk 4]
```diff
@@ -111,6 +113,55 @@ func (s *Service) isCallerAuthorized(ctx context.Context, log logrus.FieldLogger
...
+	if (len(reqSelectors) != 0 && reqPid != 0) || (len(reqSelectors) == 0 && reqPid == 0) {
+		fmt.Printf("booop")
```
Comment (azdagron) [inline pkg/agent/api/delegatedidentity/v1/service.go:127]: Leftover debug printf?

File: pkg/agent/api/delegatedidentity/v1/service_test.go [hunk 5]
```diff
@@ -79,13 +79,40 @@ func TestSubscribeToX509SVIDs(t *testing.T) {
...
 		expectResp    *delegatedidentityv1.SubscribeToX509SVIDsResponse
+		req           *delegatedidentityv1.SubscribeToX509SVIDsRequest
 	}{
 		{
 			testName:   "attest error",
 			attestErr:  errors.New("ohno"),
 			expectCode: codes.Internal,
 			expectMsg:  "workload attestation failed",
 		},
+		{
```
Comment (azdagron) [inline pkg/agent/api/delegatedidentity/v1/service_test.go:90]: I think we're missing a "happy path" test where a pid is provided and the delegate receives an X509-SVID.

File: pkg/agent/api/delegatedidentity/v1/service_test.go [hunk 6]
```diff
@@ -427,6 +457,30 @@ func TestFetchJWTSVIDs(t *testing.T) {
...
```
Comment (azdagron) [inline pkg/agent/api/delegatedidentity/v1/service_test.go:460]: I think we're missing a "happy path" test where a pid is provided and the delegate receives a JWT-SVID.

File: pkg/agent/api/delegatedidentity/v1/service_test.go [hunk 7]
```diff
@@ -735,10 +795,19 @@ type FakeAttestor struct {
 	err       error
 }
 
+type FakeWorkloadPIDAttestor struct {
```
Comment (azdagron) [inline pkg/agent/api/delegatedidentity/v1/service_test.go:798]: nit: this should be moved before the FakeAttestor.Attest method to be alongside its own methods.

**Review by azdagron — COMMENTED** [review]
Looks pretty straightforward. Just missing a few test cases. Thanks for the work, @bleggett!

**Comment by azdagron** [comment]
Hey @bleggett! Just a friendly ping about addressing the comments. I think we're pretty close here!

File: pkg/agent/api/delegatedidentity/v1/service.go (hunk 3 above)
Comment (bleggett) [reply pkg/agent/api/delegatedidentity/v1/service.go:125]: YEP. my bad

**Comment by bleggett** [comment]
@azdagron everything should be addressed, lmk if you have further comments, ty!

**Review by azdagron — APPROVED** [review]
\o/

**Comment by bleggett** [comment]
ty all!

**Pull request merged**